src/java/org/apache/cassandra/index/sai/utils/IndexFileUtils.java

```diff
@@ -59,25 +62,25 @@ public class IndexFileUtils
 
     private static final SequentialWriterOption writerOption = defaultWriterOption;
 
+    /**
+     * Remembers checksums of files so we don't have to recompute them from the beginning of the file whenever appending
+     * to a file. Keeps checksums with respective file lengths so we can detect some file changes like truncation
+     * or external appends which could omit this code, so this is still safe.
+     * Index files are append-only (immutable), so length checking is enough.
+     */
+    private static final Cache<String, Guard<FileChecksum>> checksumCache = Caffeine.newBuilder()
+                                                                                    .maximumSize(4096)
+                                                                                    .build();
+
     @VisibleForTesting
     protected IndexFileUtils()
     {}
 
     public IndexOutputWriter openOutput(File file, ByteOrder order, boolean append) throws IOException
     {
         assert writerOption.finishOnClose() : "IndexOutputWriter relies on close() to sync with disk.";
-
-        var checksumWriter = new IncrementalChecksumSequentialWriter(file);
-        IndexOutputWriter indexOutputWriter = new IndexOutputWriter(checksumWriter, order);
-        if (append)
-        {
-            // Got to recalculate checksum for the file opened for append, otherwise final checksum will be wrong.
-            // Checksum verification is not happening here as it sis not guranteed that the file has the checksum/footer.
-            checksumWriter.recalculateChecksum();
```

Comment:
Initially, I was thinking of just reading the checksum value from the previous segment's footer and feeding it into current `CRC32` but `CRC32` doesn't provide an API to set the CRC value. Not sure if it's safe to use reflection in production code.

Reply:
The would require locating the beginning of the footer first; which may be non-trivial, because it doesn't seem to be fixed-length.

Reply:
We could backwards-search for the footer magic, but there is non-zero chance that a random sequence of bytes matches the magic.

Reply:
I addressed this concern by simply computing a checksum of the footer, so we don't need to change this code if the footer ever changes.